cpp/tests/experimental/spatial/linestring_intersection_test.cu

```diff
@@ -0,0 +1,447 @@
+}
+
+TYPED_TEST(LinestringIntersectionTest, SingletoSingleOnePair)
+{
+  using T = TypeParam;
+  using P = vec_2d<T>;
+
+  using index_t = typename linestring_intersection_result<T, std::size_t>::index_t;
+  using types_t = typename linestring_intersection_result<T, std::size_t>::types_t;
+
+  auto multilinestrings1 = make_multilinestring_array({0, 1}, {0, 2}, {P{0.0, 0.0}, P{1.0, 1.0}});
+
+  auto multilinestrings2 = make_multilinestring_array({0, 1}, {0, 2}, {P{0.0, 1.0}, P{1.0, 0.0}});
+
+  auto expected = make_linestring_intersection_result<T, index_t, types_t>(
+    {0, 1}, {0}, {0}, {P{0.5, 0.5}}, {}, {0}, {0}, {0}, {0}, this->stream(), this->mr());
+
+  CUSPATIAL_RUN_TEST(this->template run_single_test<index_t>,
+                     multilinestrings1.range(),
+                     multilinestrings2.range(),
+                     expected);
+}
+
+TYPED_TEST(LinestringIntersectionTest, SingletoSingleOnePairWithDuplicatePoint)
+{
+  using T = TypeParam;
+  using P = vec_2d<T>;
+
+  using index_t = typename linestring_intersection_result<T, std::size_t>::index_t;
+  using types_t = typename linestring_intersection_result<T, std::size_t>::types_t;
+
+  auto multilinestrings1 = make_multilinestring_array({0, 1}, {0, 2}, {P{0.0, 0.0}, P{1.0, 1.0}});
+
+  auto multilinestrings2 =
+    make_multilinestring_array({0, 1}, {0, 4}, {P{0.0, 1.0}, P{1.0, 0.0}, P{0.5, 0.0}, P{0.5, 1}});
+
+  auto expected = make_linestring_intersection_result<T, index_t, types_t>(
+    {0, 1}, {0}, {0}, {P{0.5, 0.5}}, {}, {0}, {0}, {0}, {0}, this->stream(), this->mr());
+
+  CUSPATIAL_RUN_TEST(this->template run_single_test<index_t>,
+                     multilinestrings1.range(),
+                     multilinestrings2.range(),
+                     expected);
+}
+
+TYPED_TEST(LinestringIntersectionTest, SingletoSingleOnePairWithMergeableSegment)
+{
+  using T = TypeParam;
+  using P = vec_2d<T>;
+
+  using index_t = typename linestring_intersection_result<T, std::size_t>::index_t;
+  using types_t = typename linestring_intersection_result<T, std::size_t>::types_t;
+
+  auto multilinestrings1 = make_multilinestring_array({0, 1}, {0, 2}, {P{0.0, 0.0}, P{3.0, 3.0}});
+
+  auto multilinestrings2 = make_multilinestring_array(
+    {0, 1}, {0, 4}, {P{0.0, 0.0}, P{1.0, 1.0}, P{0.5, 0.5}, P{1.5, 1.5}});
+
+  auto expected =
+    make_linestring_intersection_result<T, index_t, types_t>({0, 1},
+                                                             {1},
+                                                             {0},
+                                                             {},
+                                                             {segment<T>{P{0.0, 0.0}, P{1.5, 1.5}}},
+                                                             {0},
+                                                             {0},
+                                                             {0},
+                                                             {0},
```

Comment:
The `ids` don't seem correct to me here, as they're all 0. `lhs` looks right, there's only one linestring and only one segment. On `rhs` however I'm confused at the output. There's only one linestring so 0 is correct for line #203, but line 204 is confusing. All three segments overlap, which you've combined into the result segment, but only segment 0 is returned.

If the user wanted for example to display each intersecting segment from `rhs` in different colors, this output would not enable that, I think? I'm ok with leaving the API as it is but this might need consideration.

Reply:
I think the original requirement for intersection is a ambiguous and this is certainly not considered in original design. Can file a separate issue to track if desired.

Reply:
To support your use case, you would need to group all merged geometry ids, this requires an additional level of offsets into the id arrays. I think that's a pretty big change and would take up a separate effort.

From a different angle, can't we pose this as a separate step from intersection? You have a intersection, finding where the intersection comes from is a search problem.

Reply:
We may want to factor out the intersection metadata (segment indices) as a search problem in the future anyway.

Reply:
I'm in favor of creating a new issue to track this, it doesn't need to block this PR.